### PR TITLE
PoC all own props are observable, all functions are autoAction/flow

### PR DESCRIPTION
**Pull request by urugator**
#### Changes
`deepEnhancer` automatically converts functions into `autoAction` and generators into `flow`
All own props are by default observable (like pre V6).
All explicitely provided annotations works the same as before - eg `@action` still results in non-writable non-observable action.

#### Consequences
Functions in **deep** objects/arrays/maps etc become `autoAction`s/`flow`s.
Object with **deep: false** won't turn **own** methods in `actions`/`flows`.
Since **own** methods (`autoAction`) are now observable, they're also enumerable (like pre V6).
Hopefully a bit faster `observable/makeAutoObservable`, because annotation inference is no longer a seperate step.

#### `deep: false` does not convert functions to auto actions, which is a BC (for V6), some alternatives:

a) Don't use `deepEnhancer` for conversion. Create `autoAction`/`flow` only during object initialization. Meaning the user is responsible for wrapping value in `action` or `flow` manually on assigments. Not ideal, but quite reasonable. Also it won't turn functions to autoAction/flow in other places like array etc.

b) For objects with `deep: false` use special `objectRefEnhancer`, which ignores everything, but functions. Meaning the user must use `observable.ref` explicitely for fields that shouldn't convert functions to `autoAction` even with `deep: false`. A bit weird.

#### Implementation
Introduced new (only internal) `autoAnnotation` that is used by default. It basically merges behavior of all annotations into one.

### CODE REVIEW

**Comment by mweststrate** [comment]
I don't find `c) make this a breaking change` not unreasonable either :)

File: .changeset/tender-forks-trade.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+---
+"mobx": patch
+---
+
+-   by default all functions are converted to `autoAction`s/`flow`s
+-   by default all originally _own_ props are now observable and enumerable (as in pre v6)
+-   by default _lone_ setters are converted to `action`s
+-   `deep: false` ignores _all_ property values (including functions that would be previously converted to `autoAction`/`flow`)
```
Comment (mweststrate) [inline .changeset/tender-forks-trade.md:8]: ... so actions will have to be marked explicitly as (auto) `action`

**Review by mweststrate — APPROVED** [review]
Except for an update to the docs, the proposed semantic changes look fine to me. The biggest reservation I have with the proposed changes is merely changing it in the first place. I'm not sure whether changing this again is worth the hassle (for MobX users already on 6). However, on the positive side, mobx 5 is still downloaded 4 times more than 6 (https://www.npmjs.com/package/mobx > versions > version history)

File: .changeset/tender-forks-trade.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+---
+"mobx": patch
+---
+
+-   by default all functions are converted to `autoAction`s/`flow`s
```
[the remark this answers is not in the document]
Comment (urugator) [reply .changeset/tender-forks-trade.md:5]: Well, this is not correct... `action` and `autoAction` works the same as before - both non-observable, non-enumerable.
The *"default annotation"*  for all own properties *including functions* **is `observable`** (not `autoAction`).
(To be more precise the default annotation is `autoAnnotation`, but it's behavior in this case is the same as `observable`)
`autoAction` as annotation is basically meaningless - it's no longer used even internally.
We do not need to explain what `autoAction` does or what a "default annotation" is, but what `observable` does with functions.
You know, forget about annotations ... imagine you're using V5 - everything is observable by default, even functions. Now we just change how deep enhancer treats functions - it wraps them in `autoAction`.

File: .changeset/tender-forks-trade.md (hunk 1 above)
Comment (urugator) [reply .changeset/tender-forks-trade.md:8]: Yes, it really means do not touch the values.
However `deep: false` is to create shallow observable "container" - imo it doesn't make much sense to use `deep: false` for "smart" objects with actions etc, not sure what would be a good motivation for doing so.
Also note `deep: false` won't affect methods on prototype (even when bound) - it affects only `observable` (own) props.
It is possible to change this - I can move the logic from enhancer to annotation, therefore conversion to `autoAction` would only occur for objects - but then you have the opposite problem - you cannot create an object that is truly "shallow" - which seems wrong - and if I am not mistaken, this is currently the case as well - `deep: false` has no effect on `autoAction` inferrence, so even if user creates *shallow* object, we will mess with it's content (if the content is function) - and I am not even sure if we handle this consistently, the behavior may actually differ for initialization and property created via assigment, not sure.

File: packages/mobx/src/types/autoannotation.ts [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                            : createAction(key.toString(), value, true)
+                    })
+                    if (!definePropertyOutcome) {
+                        return // intercepted
+                    }
+                    annotated = true
+                    break // use closest
+                } else {
+                    // non-bound
+                    defineProperty(source, key, {
```
[the remark this answers is not in the document]
Comment (urugator) [reply packages/mobx/src/types/autoannotation.ts:94]: Btw [refactored version](https://github.com/urugator/mobx/blob/4ea5e0b701ba331765b4eec61b7619ff57ffe824/packages/mobx/src/types/autoannotation.ts#L59-L66) looks much better as it actually delegates to individual annotations (which are also simplified).

**Comment by urugator** [comment]
Updated docs, added introductory text to the changelog.

**Review by mweststrate — APPROVED** [review]
LGTM! Feel free to rebase and merge

**Pull request merged**